### PR TITLE
Allow '1-infinity' in ranges

### LINKED ISSUE #4253: Ranges like 1-99999, {EVERYWHERE} and {INFINITY} are bugs waiting to happen
I don't like "amount=1...99999" even though we see it in existing code, because it's a bug laying in wait. "amount=1..." seems harder to parse and error prone (too visually similar to "amount=1").

_Originally posted by @GregoryLundberg in https://github.com/wesnoth/wesnoth/pull/4233#issuecomment-522594643_

### PR DESCRIPTION
**Pull request by stevecotton**
Fixes #4253.

### CODE REVIEW

**Comment by GregoryLundberg** [comment]
It might be a good idea to grep the mainline to upgrade uses (macros or otherwise). Add the changes to this PR.

**Comment by stevecotton** [comment]
WIP because src/map/location.cpp has its own range-parsing code

**Comment by CelticMinstrel** [comment]
Looks good... but do you think it should also support `-infinity`? Maybe there's no real use-case for that, I dunno...

**Comment by stevecotton** [comment]
Would `-infinity` be a negative number, a range with an implied start at zero, or a range with an implied start at 1?

**Comment by GregoryLundberg** [comment]
I think he means "-1 - -infinity"

If you allow just "infinity" to mean "1 - infinity" then I'd have "-infinity" mean "-1 - -infinity" 

Also, can you check that spaces are allowed, as I've used them, in an attempt to make this mess more readable?

**Comment by stevecotton** [comment]
Spaces are allowed, and handled by std::stoi. However, ranges currently don't support negative numbers at all, and I'd agree with @ai0867's https://github.com/wesnoth/wesnoth/pull/4233#issuecomment-522226590 that we shouldn't have `-` meaning both 'negative' and 'range separator'.

**Comment by GregoryLundberg** [comment]
Then I would say "-infinity" should be treated the same as "-7".

**Comment by stevecotton** [comment]
utils::parse_range("-7") will return {0,0}, because std::stoi("") throws invalid_argument.
src/map/location.cpp will parse it as -7, but that will never match an on-map coordinate

Currently, -7 will pass the schema validation. Given the above behavior, I think tightening the schema is the best approach.

**Comment by CelticMinstrel** [comment]
> Would `-infinity` be a negative number, a range with an implied start at zero, or a range with an implied start at 1?

A negative number. You'd write something like `-infinity-12` for example.

But if ranges already don't support negative numbers, then I guess it's fine how it is.

File: changelog.md [hunk 1]
```diff
@@ -8,6 +8,8 @@
    * Set up for translating the Wings of Victory campaign (PR#4265)
    * Changed the :help command's output to split over multiple lines
    * Added translatable explanations of :droid, :help and :idle's arguments
+ ### WML engine
+   * Ranges can now have upper limit "infinity" instead using a big number like "99" or "99999"
```
Comment (CelticMinstrel) [inline changelog.md]: Instead _of_!

File: data/schema/game_config.cfg [hunk 2]
```diff
@@ -72,16 +72,16 @@
...
-                    link=unsigned
-                [/element]
-            [/list]
+            value="\d+"
+        [/element]
+        [element]
+            value="\d+-\d+"
+        [/element]
+        [element]
+            value="\d+-infinity"
```
Comment (CelticMinstrel) [inline data/schema/game_config.cfg]: ...wait, aren't those first two `[element]` tags identical to the previously-existing `[element]` tag? Why'd you have to split it up like that?

Your commit message implies the schema was accepting negative numbers, but if the code on the left accepts negative numbers, I think that might be a bug in the schema implementation... the only way I could see it accepting negative numbers is if `unsigned` accepts an empty string.

File: changelog.md (hunk 1 above)
Comment (stevecotton) [reply changelog.md]: Oops - thanks. Will fix on Sunday, unless someone else fixes it first.

Comment (jostephd) [reply changelog.md]: Fixed in 27209171aff16a8033df4b42ac75c33e902ad425.

File: data/schema/game_config.cfg (hunk 2 above)
Comment (stevecotton) [reply data/schema/game_config.cfg]: IIUC, `[list]` accepts empty elements, as long as the number of elements that match is between min and max. So "-7" is split to "", "7", which gets one match and passes the test.

https://github.com/wesnoth/wesnoth/blob/b08677c67807a6ff2db34fcb5ebafb7f35d8ad36/src/serialization/schema/type.cpp#L115-L116

I tried changing that "true" to "false", and the validation complained about the blank lines not matching map_data in:
https://github.com/wesnoth/wesnoth/blob/b08677c67807a6ff2db34fcb5ebafb7f35d8ad36/data/scenario-test.cfg#L3127-L3138

Comment (CelticMinstrel) [reply data/schema/game_config.cfg]: Hmm... I'm not sure what's best here... obviously the map data needs to match, but I don't think skipping empty lines is even quite the right approach there (is a blank line in the middle of the map actually valid?)... and I think for most lists not splitting on line breaks, empty elements are undesirable.

**Pull request merged**